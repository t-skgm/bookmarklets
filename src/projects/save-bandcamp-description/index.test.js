import { jest } from '@jest/globals';
import { JSDOM } from 'jsdom';
import { readFileSync } from 'fs';
import { resolve } from 'path';
// import { setTimeout } from 'timers/promises';
import { buildHtmlWithJsonLd, getDirname, mockedFetchFn } from '../../../test/utils';

const stubHtml = readFileSync(resolve(getDirname(import.meta.url), './__stub__/bandcamp_album.html'), {
  encoding: 'utf8'
});

describe('save-bandcamp-description', () => {
  /** @type {JSDOM} */
  let dom;
  /** @type {Window} */
  let window;
  /** @type {Document} */
  let document;

  beforeAll(() => {
    // init window/document with jsdom
    dom = new JSDOM(stubHtml);
    global.document = document = dom.window.document;
    global.window = window = dom.window;
    global.DOMParser = dom.window.DOMParser;

    // mock fetch
    global.fetch = jest.fn(mockedFetchFn(buildHtmlWithJsonLd));
  });

  test('test jsdom mocking', async () => {
    const documentTitle = document.querySelector('title').textContent;
    expect(documentTitle).toMatchInlineSnapshot(`"Crazy Faces (Songs For Snod (2012-2019)) | Martin Carr"`);
  });

  describe('execute script', () => {
    let linkClickSpy;
    let createObjectURLSpy;

    beforeAll(async () => {
      // set spy fn
      linkClickSpy = jest.spyOn(window.HTMLElement.prototype, 'click');
      createObjectURLSpy = jest.spyOn(global.URL, 'createObjectURL');

      // execute script
      await import('./index');
    });

    test('download file (click link) has fired 2 times', () => {
      expect(linkClickSpy).toBeCalledTimes(1);
    });

    test('generated text data', () => {
      const textData = createObjectURLSpy.mock.calls[0][0].blobParts[0];

      expect(textData).toMatchSnapshot();
    });
  });
});
