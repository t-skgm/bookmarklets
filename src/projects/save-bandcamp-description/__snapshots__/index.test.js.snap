// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`save-bandcamp-description execute script verify generated text data 1`] = `
"Martin Carr
Crazy Faces (Songs For Snod (2012-2019))
https://martincarr.bandcamp.com/album/crazy-faces-songs-for-snod-2012-2019

<Track list>
1. One Fine Day
2. Crazy Faces
3. I'm Alright
4. Cal
5. Easy Come, Easy Go
6. World In Action
7. The Years Roll Down My Face
8. Jack Jones
9. Only You Can See Me
10. Land of the Living
11. Snodgrass
12. The Years Roll Down My Face (Easy Come, Easy Go) Reprise
13. Your Body

<Description>
Sometime ago, my friend David Quantick sent me a book of short stories, urging me to read the one called Snodgrass by Ian R Mcleod. Alternative histories don't tend to interest me too much but Snodgrass made me laugh out loud. I filed it alongside National Lampoon's 'Tragical Misery Tour' in my internal Non-Beatle Beatle things that are very funny folder. 

Fast forward to 2012 and I'm ten years adrift, earning no money and unable to finish anything. I've cut myself off from social media and, with two very young children, sleeping even less than usual. DQ sends me an email asking me to write the music for a tv adaptation of Snodgrass. Yeah! I say. Yeah! Yeah! Yeah!

I meet the producer, Mike, and the three of us have long and boozy conversations about script and music and The Beatles. I set up a slapback delay and every morning I go into the studio and pretend I'm a middle-aged man who threw it all away. Turns out there was no pretense necessary. 

Later, I catch the red-eye up to Manchester to meet up with the director David Blair, the director, Ian R McLeod, Mike and DQ, and Ian Hart, who is playing Lennon for the third time. There follows a lot of drinking and laughing. 

The film is shown on Sky Arts and that is that.

Since then there has been talk of a feature film. Songs and scripts have been written, actors auditioned, and money chased but it looks like it's not going to happen. 

These are not the versions used in the film, I don't think I own those. These are the demos, they crackle and pop but I think they sound pretty good. 

I hope you enjoy this album x

<Credits>
All songs by Martin Carr
The cover is a photograph I took during the filming, somewhere in Manchester, 2012.

<Published>
2022-03-05

<Lisence>
All Rights Reserved

<Tags>
Pop, Cardiff"
`;
